src/Microsoft.TemplateEngine.Edge/Settings/SettingsLoader.cs

```diff
@@ -289,115 +294,102 @@ private void EnsureLoaded()
+                mountPoints[package.MountPointUri] = package.LastChangeTime;
 
-                    if (_userTemplateCache.MountPointsInfo.TryGetValue(source.MountPointUri, out var cachedLastChangeTime))
+                // We can stop comparing, but we need to keep looping to fill mountPoints
+                if (!needsRebuild)
+                {
+                    if (cache.MountPointsInfo.TryGetValue(package.MountPointUri, out var cachedLastChangeTime))
                     {
-                        if (source.LastChangeTime > cachedLastChangeTime)
+                        if (package.LastChangeTime > cachedLastChangeTime)
                         {
-                            placesThatNeedScanning.Add(source.MountPointUri);
+                            needsRebuild = true;
                         }
                     }
                     else
                     {
-                        placesThatNeedScanning.Add(source.MountPointUri);
+                        needsRebuild = true;
                     }
                 }
             }
 
-            foreach (var place in mountPoints.Where(k => placesThatNeedScanning.Contains(k.Key)).OrderBy(k => k.Value))
+            // Check that some mountpoint wasn't removed...
+            if (!needsRebuild && mountPoints.Keys.Count != cache.MountPointsInfo.Count)
+            {
+                needsRebuild = true;
+            }
+
+            // Cool, looks like everything is up to date, exit
+            if (!needsRebuild)
+            {
+                return cache;
+            }
+
+            var scanResults = new List<(ITemplatePackage Package, ScanResult ScanResult)>();
+            Parallel.ForEach(allTemplatePackages, (package) =>
             {
                 try
                 {
-                    _userTemplateCache.Scan(place.Key);
+                    var scanResult = _installScanner.Scan(package.MountPointUri);
+                    lock (scanResults)
+                    {
+                        scanResults.Add((package, scanResult));
+                    }
                 }
                 catch (Exception ex)
                 {
-                    _environmentSettings.Host.OnNonCriticalError(null, $"Failed to scan \"{place}\":{Environment.NewLine}{ex}", null, 0);
+                    _environmentSettings.Host.OnNonCriticalError(null, $"Failed to scan \"{package.MountPointUri}\":{Environment.NewLine}{ex}", null, 0);
                 }
-            }
-
-            // When writing the template caches, we need the existing cache version to read the existing caches for before updating.
-            // so don't update it until after the template caches are written.
-            _userTemplateCache.WriteTemplateCaches(mountPoints);
+            });
+
+            cache = new TemplateCache(
+                scanResults.OrderBy((p) => (p.Package.Provider.Factory as IPrioritizedComponent)?.Priority).Select((pair) => pair.ScanResult),
```

Comment:
Once this is ordered, null values will come before the negative ones. Is that the desired behavior? Here is a sample ordering:
```
null. null. null, -998, -98, -5, 0, 3, 7, 999
```

If not, maybe we should do this instead:
```suggestion
                scanResults.OrderBy((p) => (p.Package.Provider.Factory as IPrioritizedComponent)?.Priority ?? 0).Select((pair) => pair.ScanResult),
```

Reply:
Thank you!